### DISCUSSION TITLE
How to modify pull request body without clearing it on update

### DISCUSSION THREAD

**Discussion by realloser** [renovatebot/renovate#18914] [Request Help]
This is related to this [question](https://github.com/renovatebot/renovate/discussions/17888) and [issue](https://github.com/renovatebot/renovate/issues/17897).

Is there a way to prevent RenovateBot from removing changes/additions in the pull request body (e.g. notes) when a pull request gets updated by renovate.

While this [issue](https://github.com/renovatebot/renovate/issues/17897) is still collecting the requirements. I would argue, that it should be possible to add notes to a pull request body and make sure that the notes aren't deleted. Therefore the renovate related body should be tagged in a way to make sure, only the tagged section is replaced upon an update of the pull request.
This would:
1. allow adding notes for a PR without losing them
2. manually add the PR-Template and be sure, it doesn't get cleared (as a first step and workaround until the final [issue](https://github.com/renovatebot/renovate/issues/17897) has been addressed)

Sample: 
```
<!--renovate:start-->
renovate body
<!--renovate:end-->
```

if the tags are missing or incomplete, the full body is replaced. I would assume this would be an "easy" enhancement?

**Comment by rarkins** [comment]
Not all platforms Renovate supports can themselves support hidden html comments. e.g. GitHub and GitLab can.

But this would be OK if we can gracefully "degrade" on those which don't support it, so that functionality remains as it was.

Would you be interested in contributing this enhancement if so?

**Reply by realloser** [reply]
I didn't think of other platform. Good to know. 

If you think this would be a good starting point for a first contribution (complexity, ...), I definitely would be interested to bring this forward, as it is currently blocking/slowing down our workflow.

**Comment by PhilipAbed** [comment]
@realloser you can create an issue and handle it if you want, it's not that hard, 
we already have a similar hidden comment for github/gitlab , you can follow changes depending on where it is being handled

First Destructing the PR:
https://github.com/renovatebot/renovate/blob/e4e34ff73f046029735a6482154feeb43afb240d/lib/modules/platform/pr-body.ts#L74-L81

While constructing new body, add your saved comment back:
https://github.com/renovatebot/renovate/blob/e4e34ff73f046029735a6482154feeb43afb240d/lib/workers/repository/update/pr/body/index.ts#L99-L100

Then massage body in other platforms: in this method: `platform.massageMarkdown(prBody);`
see example:
https://github.com/renovatebot/renovate/blob/e4e34ff73f046029735a6482154feeb43afb240d/lib/modules/platform/bitbucket/index.ts#L479-L491

enjoy

**Reply by realloser** [reply]
Nice. Thank you for the pointers.

**Comment by realloser** [comment]
@viceice @rarkins [This PR#19596](https://github.com/renovatebot/renovate/pull/19596) tries to addesse the issue discussed here. 
Seems as we are hitting a problem with the caching, which is causing a problem as we don't know whether the PR body has changed. There are events (update) and I'm not that familiar with the implementation and help would be appreciated.
Do you have any solution in mind to get around the caching problem? 
Looking at the logic some of the implementations do pass in flags to disable caching for a given PR on update in case of an event. Would that be a possible solution? We could then also use `etag` to have reduced payload in case the cache is up to date. But my assumption is, that in case of an event, `etag` is outdated.

**Reply by rarkins** [reply]
> Seems as we are hitting a problem with the caching, which is causing a problem as we don't know whether the PR body has changed.

Can you elaborate on this? I'm not sure it's a cache problem (as in.. cache invalidation), I thought we had a "Renovate only stores a hash of the contents and not the full contents" problem.

**Reply by realloser** [reply]
It is correct. Renovate bot only caches the hash of the contents.
The goal of this "improvement" is to make sure we only update the renovate specific body in a PR body. To do so we have to get the PR body. Using the cache now renders obsolete as we need the full body for the update. The has is only used to verify whether the new renovate body hash would be different to the cached renovate body hash. If this is the case it will update the pr body with the new renovate body, which is not desired as we want to keep the non renovate body specific changes.

Maybe I'm just missing a peace?

**Reply by viceice** [reply]
we somewhere discussed that we need to update platform code, so we can fetch full PR body if we detect a body hash mismatch.

I think there's a open PR for it. we don't like to cache the full body because of size

**Reply by realloser** [reply]
Caching is not an option and a requirement anyway, agree on that.
There are two important usecases I think:
- PR Body gets never updated
- PR Body gets updated (e.g. PR-Template, checkboxes click, text)

fFor the second uscase we have to pull data from Github on every PR update. I think, we cannot distinct, whether the body has updated or a comment was added or any other unrelated change.
If we add downloading latest PR Data, we would „punish“ the users of use case 1 with unnecessary „expensive“ GH calls, since the body never changes. Is this acceptable?
Should we invalidate the cache and pull again the data on every update?
Should we have a configuration for that to allow the second use case without punishing scenario 1?
We can improve calls using GH GraphQL and reduce the return to the body only?

Back to a question in the PR. Renovate Body consists of a checkbox to rebase. How is this evaluated. How does Renovate knows, that this checkbox was set? I assume this would require the body to be downloaded and parsee. If you can point me to this logic, I may be able to use this. I‘m not even sure whether this works at all looking at the code and the caching 🤷🏻‍♂️.

**Reply by rarkins** [reply]
Check this interface out: https://github.com/renovatebot/renovate/blob/05f0a499b8a16dfe67c37d074be045d50ee1f169/lib/modules/platform/types.ts#L52-L57

`hash` refers to the has of the PR body content, but not the raw content. For example `rebaseRequested` indicates whether the checkbox is ticked or not.

I think you could add to that two string fields - `preContent` and `postContent`. This would include any raw PR body content which has been added before or after Renovate's content. Then you could reconstruct the PR by combining preContent, then Renovate's normal content, then postContent.

**Reply by realloser** [reply]
thanks for the input! I have found ˋrebaseRequestedˋ but 8 could‘t figure out how this is set on an update. Any idea on how this happens for Github without fetching the body? For me it looks like for Github Renovate always works with the cache. If this is true, this flag will never be true.

**Reply by rarkins** [reply]
Renovate always fetches the full PR body, but it _saves_ only the info it needs (such as the hash of content, not the full content). See https://github.com/renovatebot/renovate/blob/05f0a499b8a16dfe67c37d074be045d50ee1f169/lib/modules/platform/pr-body.ts#L56

**Reply by realloser** [reply]
Yes, and it does it as long the PR is not chached. Correct? In case of an GH update I could find logic which fechtes it again on the event of an update.

**Reply by rarkins** [reply]
Caching is "busted" whenever the PR gets modified between runs
